contracts/Dex.sol

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.6.5;
+
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+contract Dex is Initializable, AccessControlUpgradeable {
+    using SafeMath for uint256;
+
+    bytes32 public constant GAME_ADMIN = keccak256("GAME_ADMIN");
+    uint public constant FEE_DENOMINATOR = 10000;
+
+    uint8 public constant VAR_FEE = 0;
+
+    struct TokenPair {
+        address token1;
+        uint token1Amount;
+        address token2;
+        uint token2Amount;
+    }
+
+    mapping(uint => TokenPair) public tokenPairs;
+    mapping(uint256 => uint256) public vars;
+
+    event TokenPairSwapped(address indexed token1, uint token1Amount, address indexed token2, uint token2Amount);
+    event TokenPairAdded(uint id, address indexed token1, address indexed token2);
+    event LiquidityAdded(address indexed token1, uint token1Amount, address indexed token2, uint token2Amount);
+
+    function initialize() virtual public initializer {
+        __AccessControl_init_unchained();
+        _setupRole(DEFAULT_ADMIN_ROLE, tx.origin);
+        _setupRole(GAME_ADMIN, tx.origin);
+    }
+
+    modifier restricted() {
+        _restricted();
+        _;
+    }
+
+    function _restricted() internal view {
+        require(hasRole(GAME_ADMIN, msg.sender) || hasRole(DEFAULT_ADMIN_ROLE, msg.sender), "NA");
+    }
+
+    // FUNCTIONS
+
+    function swap(address tokenA, address tokenB, uint amountA) external {
+        uint id = getTokenPairId(tokenA, tokenB);
+        TokenPair memory pair = tokenPairs[id];
+        require((pair.token1 == tokenA && pair.token2 == tokenB) || (pair.token2 == tokenA && pair.token1 == tokenB), "Invalid pair");
+        uint amountB = getAmountOut(tokenA, tokenB, amountA);
+        require(amountB > 0, "Invalid amount");
+        if (tokenA == pair.token1) {
+            pair.token1Amount += amountA;
+            pair.token2Amount -= amountB;
```

Comment:
You should update amount in pair with the same amount you're using for transfers.
So in this case `amountB - amountB.mul(vars[VAR_FEE]).div(FEE_DENOMINATOR)` instead of just `amountB`.
Sam in line 58.